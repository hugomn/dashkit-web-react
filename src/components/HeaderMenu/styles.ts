import { Box, Button, Drop, Text } from "grommet";
import styled from "styled-components";
import theme from "../../constants/theme";

export const StyledButton = styled(Button)`
  /* font-weight: bold; */
  &:hover {
    /* background-color: ${theme.global.colors["dark-3"]}; */
    /* color: white; */
  }
`;

export const StyledText = styled(Text)`
  transition: all 0.3s ease-in-out;
`;

export const IconWrapper = styled(Box)<{ inverted?: boolean }>`
  transition: all 0.3s ease-in-out;
  transform: ${({ inverted }) => (inverted ? "rotate(-180deg)" : "rotate(0deg)")};
`;

export const StyledDrop = styled(Drop)`
  margin-top: 24px;
  overflow: visible;
  &::after {
    content: "";
    position: absolute;
    top: 10px;
    left: 50%;
    margin-top: -16px;
    border-width: 6px;
    border-style: solid;
    border-color: ${theme.global.colors.white} ${theme.global.colors.white} transparent transparent;
    transform: rotate(-45deg);
    box-shadow: 1px -2px 2px 0px rgba(0, 0, 0, 0.1);
  }
`;

export const ContainerBox = styled(Box)`
  max-height: inherit;
  /* IE11 hack to get drop contents to not overflow */
  @media screen and (-ms-high-contrast: active), (-ms-high-contrast: none) {
    width: 100%;
  }
  ${props => props.theme.menu.extend};
`;
