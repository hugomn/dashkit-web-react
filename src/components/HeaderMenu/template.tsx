import { Box, Button } from "grommet";
import { Down } from "grommet-icons";
import React, { Dispatch, RefObject, SetStateAction, useRef, useState } from "react";
import { ContainerBox, IconWrapper, StyledButton, StyledDrop, StyledText } from "./styles";
import { IItem, IProps } from ".";

export const HeaderMenu: React.FC<IProps> = ({ active, id, items, label, margin, open }) => {
  const ref: RefObject<any> = useRef();
  const [show, setShow] = useState(open || false);
  return (
    <Box align="start" justify="center" margin={margin}>
      <Button id={id} ref={ref} onClick={toggleShow(show, setShow)}>
        <Box direction="row" justify="start" align="center" pad="small" gap="small">
          <StyledText size="medium" weight="bold" color={active || show ? "brand" : ""}>
            {label}
          </StyledText>
          <IconWrapper inverted={show}>
            <Down color={active || show ? "brand" : ""} size="small" />
          </IconWrapper>
        </Box>
      </Button>
      {show && (
        <StyledDrop
          id={id ? `${id}__drop` : undefined}
          align={{ top: "bottom" }}
          target={ref.current}
          onClickOutside={handleClickOutside(setShow)}
          elevation="small">
          <ContainerBox background="white">
            <Box overflow="auto">{items.map(renderItem(setShow))}</Box>
          </ContainerBox>
        </StyledDrop>
      )}
    </Box>
  );
};

// comment

// eslint-disable-next-line react/display-name
export const renderItem = (setShow: Dispatch<SetStateAction<boolean>>) => (
  item: IItem,
  index: number
) => (
  <Box key={index} flex={false} data-testid="header-item">
    <StyledButton
      hoverIndicator="background"
      {...{ ...item, icon: undefined, label: undefined }}
      onClick={handleItemClick(setShow, item.onClick)}>
      <Box align="start" pad="small" direction="row">
        {item.icon}
        {item.label}
      </Box>
    </StyledButton>
  </Box>
);

export const handleItemClick = (
  setShow: Dispatch<SetStateAction<boolean>>,
  onClick?: (...args: any[]) => void
) => (...args: any[]) => {
  if (onClick) {
    onClick(...args);
  }
  setShow(false);
};

export const handleClickOutside = (setShow: Dispatch<SetStateAction<boolean>>) => () => {
  setShow(false);
};

export const toggleShow = (show: boolean, setShow: Dispatch<SetStateAction<boolean>>) => () => {
  setShow(!show);
};

export default HeaderMenu;
