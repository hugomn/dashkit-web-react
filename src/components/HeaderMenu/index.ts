import { BoxProps } from "grommet";
import template from "./template";

export interface IItem {
  icon?: string;
  label: string;
  onClick?: (...args: any[]) => void;
}

export interface IProps {
  active?: boolean;
  id?: string;
  items: IItem[];
  label: string;
  margin?: BoxProps["margin"];
  open?: boolean;
}

export default template;
