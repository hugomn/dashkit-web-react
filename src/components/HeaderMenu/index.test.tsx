import { getByText as getByTextDOM } from "@testing-library/dom";
import { cleanup, fireEvent, render } from "@testing-library/react";
import * as React from "react";
import HeaderMenu, {
  handleClickOutside,
  handleItemClick,
  renderItem,
  toggleShow
} from "./template";

describe("HeaderMenu component", () => {
  afterEach(cleanup);

  it("hides the drop when clicking an item", () => {
    const onClick = jest.fn();
    const { getByText, container } = render(
      <HeaderMenu
        id="test-menu"
        label="Test"
        items={[{ label: "Item 1", onClick }, { label: "Item 2" }]}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
    fireEvent.click(getByText("Test"));
    fireEvent.click(getByTextDOM((document as unknown) as HTMLElement, "Item 1"));
    expect(onClick).toBeCalled();
    expect(document.getElementById("test-menu__drop")).toBeNull();
  });

  it("render correctly without an id", () => {
    const { getByText } = render(<HeaderMenu label="Test" items={[{ label: "Item 1" }]} />);
    fireEvent.click(getByText("Test"));
    expect(getByText("Test")).toBeTruthy();
  });

  it("Should render an item correctly", () => {
    const mockSetShow = jest.fn();
    const { getByText } = render(renderItem(mockSetShow)({ label: "Test item" }, 0));
    expect(getByText("Test item")).toBeTruthy();
  });

  describe("handleItemClick", () => {
    it("Should correctly close the menu ", () => {
      const mockSetShow = jest.fn();
      const mockArg = { arg: "test" };
      handleItemClick(mockSetShow)(mockArg);
      expect(mockSetShow).toBeCalledWith(false);
    });

    it("Should correctly call the onClick callback", () => {
      const mockOnClick = jest.fn();
      const mockArg = { arg: "test" };
      handleItemClick(jest.fn(), mockOnClick)(mockArg);
      expect(mockOnClick).toBeCalledWith(mockArg);
    });
  });

  describe("handleClickOutside", () => {
    it("Should close the menu", () => {
      const mockSetShow = jest.fn();
      handleClickOutside(mockSetShow)();
      expect(mockSetShow).toBeCalledWith(false);
    });
  });

  describe("toggleShow", () => {
    it("Should close the menu", () => {
      const mockSetShow = jest.fn();
      toggleShow(true, mockSetShow)();
      expect(mockSetShow).toBeCalledWith(false);
    });

    it("Should open the menu", () => {
      const mockSetShow = jest.fn();
      toggleShow(false, mockSetShow)();
      expect(mockSetShow).toBeCalledWith(true);
    });
  });
});
