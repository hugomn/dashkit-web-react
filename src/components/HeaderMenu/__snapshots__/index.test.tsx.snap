// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderMenu component hides the drop when clicking an item 1`] = `
<div
  class="StyledBox-sc-13pk1d4-0 etUtOO"
>
  <button
    class="StyledButton-sc-323bzc-0 jfbKsv"
    id="test-menu"
    type="button"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 lgRCHI"
    >
      <span
        class="sc-bwzfXH cJpiqs StyledText-sc-1sadyjn-0 bnqisW"
      >
        Test
      </span>
      <div
        class="StyledBox__StyledBoxGap-sc-13pk1d4-1 iChEkS"
      />
      <div
        class="sc-htpNat fmorob StyledBox-sc-13pk1d4-0 IRSNj"
      >
        <svg
          aria-label="Down"
          class="StyledIcon-ofa7kd-0 fjELgu"
          viewBox="0 0 24 24"
        >
          <polyline
            fill="none"
            points="7.086 3.174 17.086 13.174 7.086 23.174"
            stroke="#000"
            stroke-width="2"
            transform="scale(1 -1) rotate(-89 -1.32 0)"
          />
        </svg>
      </div>
    </div>
  </button>
</div>
`;
